src/android/com/eclipsesource/firebase/reference/Reference.java

```diff
@@ -0,0 +1,182 @@
+        Reference ref = new Reference(this.activity, this.tabrisContext);
+        ref.create(this.database.push());
+        return ref;
+    }
+
+    public void keepSynced(boolean keepSynced){
+        this.database.keepSynced(keepSynced);
+    }
+
+    public void setValue(Object value){
+        //Set the value and add a CompletitionListener
+        database.setValue(value, completition);
+    }
+
+    public String getKey(){
+        return this.database.getKey();
+    }
+
+
+
+
+    private void registerListeners(){
+        database.addValueEventListener(valueListener);   
+
+        database.addChildEventListener(childlistener);
+    }
+    
+    private DatabaseReference.CompletionListener completition = new DatabaseReference.CompletionListener() {
+        @Override
+        public void onComplete(DatabaseError error, DatabaseReference ref){
+            Log.e("Nery", "Saved");
+            final RemoteObject remoteObject = tabrisContext.getObjectRegistry().getRemoteObjectForObject(Reference.this);
+            if (remoteObject != null) {
+                if(error != null){
+                    //... notify if there was an error.
+                    remoteObject.notify("ErrorSaving", "error", error.getMessage());
+                } else {
+                    //... if everything OK, 
+                    ref.addListenerForSingleValueEvent(new ValueEventListener() {
+                        @Override
+                        public void onDataChange(DataSnapshot dataSnapshot) {
+                            // notify that tha value was saved.
+                            remoteObject.notify("onValueSaved", "value", dataSnapshot.getValue());
+                        }
+                        @Override public void onCancelled(DatabaseError databaseError) {}
+                    });
+                }
+            }
+        }
+    };
+
+    private ValueEventListener valueListener = new ValueEventListener() {
+        @Override
+        public void onDataChange(final DataSnapshot dataSnapshot) {
+            Log.e("Nery", "Changed..!");
+            if(activity == null){
+                Log.e("ERROR", "activity es null");
+            }
+            if(tabrisContext == null){
+                Log.e("ERROR", "tabrisContext es null");
+            }
+            RemoteObject remoteObject = tabrisContext.getObjectRegistry().getRemoteObjectForObject(Reference.this);
+            if (remoteObject != null) remoteObject.notify("onDataChange", "data", dataSnapshot.getValue());
+        }
+        @Override public void onCancelled(DatabaseError databaseError) {}
+    };
+    
+    private ChildEventListener childlistener = new ChildEventListener(){
+        @Override
+        public void onCancelled(DatabaseError error){
+            RemoteObject remoteObject = tabrisContext.getObjectRegistry().getRemoteObjectForObject(Reference.this);
+            if (remoteObject != null) remoteObject.notify("onCancelled", "error", error.getMessage());
```

Comment:
The protocol event names should not contain the "on" prefix as they are registered in an `ref.on('cancelled',  () => ...)` call anyways.